intel_extension_for_transformers/llm/library/jblas/jblas/jit_blas_device.h

```diff
@@ -252,12 +370,96 @@ class CpuDevice {
   }
 #undef ADD_FLAG
 
+  static bool isEcore() {
+    Xbyak::util::Cpu cpu;
+    uint32_t tmp[4];
+    cpu.getCpuid(0x1A, tmp);
+    int core_type = (tmp[0] >> 24) & ((1u << 7) - 1);  // cpu.extractBit(a[0], 24, 31);
+    switch (core_type) {
+      case 32:
+        // printf("Atom\n");
+        return true;  // E-core or LPE-core
+        break;
+      case 64:
+        // printf("Core\n");
+        return false;  // P-core
+        break;
+      default:
+        // printf("No hyper\n");
+        return false;
+        break;
+    }
+    return false;
+  }
+
+  int getCoreId(Xbyak::util::Cpu& cpu) {
+    uint32_t tmp[4];
+    cpu.getCpuidEx(0x1F, 1, tmp);  // sub-leaf 1 is core domain
+    // printf("!!!%x\t%x\t%x\t%x!!!\n", tmp[0], tmp[1], tmp[2], tmp[3]);
+    if (tmp[0] != 0 && tmp[1] != 0)
+      return tmp[3] >> 3;  // tmp[3] is APIC
+    else
+      return tmp[3];
+  }
+
+  bool isEcore(Xbyak::util::Cpu& cpu) {
+    uint32_t tmp[4];
+    cpu.getCpuid(0x1A, tmp);
+    int core_type = (tmp[0] >> 24) & ((1u << 7) - 1);  // cpu.extractBit(a[0], 24, 31);
+    switch (core_type) {
+      case 32:
+        // printf("Atom\n");
+        return true;  // E-core or LPE-core
+        break;
+      case 64:
+        // printf("Core\n");
+        return false;  // P-core
+        break;
+      default:
+        // printf("No hyper\n");
+        return false;
+        break;
+    }
+    return false;
+  }
+  static void core_bond(int core) {
+#ifdef _WIN32
+    SetThreadAffinityMask(GetCurrentThread(), 1 << core);
+#else
+    cpu_set_t cpuset;
+    CPU_ZERO(&cpuset);
+    CPU_SET(core, &cpuset);
+    int s = sched_setaffinity(0, sizeof(cpu_set_t), &cpuset);
+    if (s != 0) printf("ERROR\n");
+#endif
+  }
+
+  static void core_bond(std::thread& thread, int core) {
+#ifdef _WIN32
+    HANDLE handle = thread.native_handle();
+    SetThreadAffinityMask(handle, 1 << core);
+#else
+    cpu_set_t cpuset;
+    CPU_ZERO(&cpuset);
+    CPU_SET(core, &cpuset);
+    pthread_t pt = thread.native_handle();
+    int s = pthread_setaffinity_np(pt, sizeof(cpuset), &cpuset);
+    if (s != 0) printf("ERROR\n");
+#endif
+  }
+
+  bool isHybrid() { return mHybrid; }
+
  protected:
   uint32_t L2Cache, L1Cache;
+  bool mHybrid = false;
   bool mHasAVX2, mHasAVX_VNNI, mHasAVX, mHasAVX512_VNNI, mHasAMX_INT8, mHasAMX_BF16, mHasAVX512F, mHasAVX512_BF16,
       mHasAVX512_FP16;
   int numcores;
   int numthreads;
+  std::vector<int> P_core, E_core, smt_core;
+  uint32_t E_L2Cache, E_L1Cache;
+  float P_power = 3.8, E_power = 2.7;
```

Comment:
Just curious: why do we choose these two const values? Is the P/E core scheduling performance determined by these two const values also perform best in future newer architectures? If not, is there an opportunity to make it tunable?

Reply:
I think this is all core frequency. we might need to configure this generation by generation

Reply:
It's a good problem. This value is measured on our MTL device. They should be measured by device, especially for laptop. Different manufacturers have different performance strategies. We have get two P/E scales (1.1 1.3) on the same CPU from the third-party reviews now. It has not a good solution now. We perhaps tune it before inference or fix a scale CPU by CPU. But I think it is not the first problem to solve in this PR. We can discuss it later.